File: gcc/testsuite/c-c++-common/gomp/target-map-iterators-6.c
/* { dg-do compile } */
/* { dg-options "-fopenmp -fdump-tree-gimple" } */

#define DIM 17

void f (int **x)
{
    #pragma omp target data map(iterator(it = 0:DIM), tofrom: x[it])
      {
        for (int i = 0; i < DIM; i++)
          *x[i] += i;
      }
}

/* { dg-final { scan-tree-dump "if \\(i <= 16\\) goto <D\\\.\[0-9\]+>; else goto <D\\\.\[0-9\]+>;" "gimple" } } */
/* { dg-final { scan-tree-dump "map\\(iterator\\(int it=0:17:1, .+\\):tofrom:\\*D\\\.\[0-9\]+" "gimple" } } */
/* { dg-final { scan-tree-dump "map\\(iterator\\(int it=0:17:1, .+\\):attach:x" "gimple" } } */

File: gcc/testsuite/gfortran.dg/gomp/target-map-iterators-6.f90
! { dg-do compile }
! { dg-options "-fopenmp -fdump-tree-omplower" }

module m
  implicit none
  integer, parameter :: DIM = 17
  type :: array_ptr
    integer, pointer :: ptr(:)
  end type
contains
  subroutine f(x)
    type (array_ptr) :: x(DIM)
    integer :: i

    !$omp target data map(iterator(it = 1:DIM), tofrom: x(it)%ptr(:))
      do i = 1, DIM
        x(i)%ptr(1) = i
      end do
    !$omp end target data
  end subroutine
end module

! { dg-final { scan-tree-dump "if \\(it <= 17\\) goto <D\\\.\[0-9\]+>; else goto <D\\\.\[0-9\]+>;" "omplower" } }
! { dg-final { scan-tree-dump "map\\(iterator\\(integer\\(kind=\[0-9\]+\\) it=1:17:1, loop_label=<D\\\.\[0-9\]+>, elems=omp_iter_data\\\.\[0-9\]+, index=D\\\.\[0-9\]+\\):to:\\(\\\*x\\\.\[0-9\]+\\)\\\[D\\\.\[0-9\]+\\\]\\\.ptr" "omplower" } }
! { dg-final { scan-tree-dump "map\\(iterator\\(integer\\(kind=\[0-9\]+\\) it=1:17:1, loop_label=<D\\\.\[0-9\]+>, elems=omp_iter_data\\\.\[0-9\]+, index=D\\\.\[0-9\]+\\):tofrom:MEM" "omplower" } }
! { dg-final { scan-tree-dump "map\\(iterator\\(integer\\(kind=\[0-9\]+\\) it=1:17:1, loop_label=<D\\\.\[0-9\]+>, elems=omp_iter_data\\\.\[0-9\]+, index=D\\\.\[0-9\]+\\):attach:\\(\\\*x\\\.\[0-9\]+\\)\\\[D\\\.\[0-9\]+\\\]\\\.ptr.data" "omplower" } }
